internal/clients/kafka/acl/acl.go

```diff
@@ -0,0 +1,194 @@
+package acl
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/crossplane-contrib/provider-kafka/apis/acl/v1alpha1"
+
+	"github.com/pkg/errors"
+	"github.com/twmb/franz-go/pkg/kadm"
+	"github.com/twmb/franz-go/pkg/kmsg"
+)
+
+// AccessControlList is a holistic representation of a Kafka ACL with configurable
+// fields
+type AccessControlList struct {
+	Name                            string
+	ResourceType                    string
+	AccessControlListPrinciple      string
+	AccessControlListHost           string
+	AccessControlListOperation      string
+	AccessControlListPermissionType string
+	ResourcePatternTypeFilter       string
+}
+
+// Topic is a holistic representation of a Kafka Topic with all configurable
+// fields
+type Topic struct {
+	Name              string
+	ReplicationFactor int16
+	Partitions        int32
+	ID                string
+	Config            map[string]*string
+}
+
+// List lists all the ACLs in Kafka
+func List(ctx context.Context, cl *kadm.Client, accessControlList *AccessControlList) (*AccessControlList, error) {
```

Comment:
Hmm, checking the terraform provider, it looks like it does not support update for ACLs, e.g. there is no update method implemented: https://github.com/Mongey/terraform-provider-kafka/blob/master/kafka/resource_kafka_acl.go

Reply:
Ideally, our resource should reject any updates after creation, simply returning an error. However, this is [not possible](https://github.com/crossplane/crossplane/issues/727) today but hopefully will be, with an admission [webhook](https://github.com/crossplane/crossplane/issues/1152) in the near future.

There are other resources in other providers that do not support updates. However, we have another challenge here that is the lack of a unique identifier. 

> We could have ResourceUpToDate set to false, which would send a user to Create when they update any configs.

Wouldn't this still cause the creation of a new resource with updated config and losing track of the previous one as above 🤔

I am wondering if we can leverage `external-name` annotation here. Building a _unique_ identifier by concatenating all the fields (similar to what terraform provider uses for [import](https://github.com/Mongey/terraform-provider-kafka/pull/152#issuecomment-728120534)) and at the beginning of observe, compare it with the new spec and reject going further with a proper error like `updates not supported` if it is different.

This would require disabling [NameAsExternalName](https://github.com/crossplane/crossplane-runtime/blob/428b7c3903756bb0dcf5330f40298e1fa0c34301/pkg/reconciler/managed/api.go#L46) initializer like [this](https://github.com/crossplane/provider-aws/blob/b3f7c2b86595f053bf9295cd05dfac5a71d1535a/pkg/controller/servicediscovery/httpnamespace/setup.go#L60) and setting it once after create.

Reply:
We've been thinking about this, and don't feel like we have a good solution.  I'm gonna @twmb to see if he has thoughts on whether this is feasible or not.  @schaubce @rscibelli too.

**Problem:**

Present state: One managed resource, say `marshmallory`, to all ACLs named `marshmallory`.  Because K8s doesn't have a unique ID, create runs, the first `marshmallory` gets orphaned, and the latest yaml is applied.  

Desired state: One managed resource, `marshmallory`, maps to a unique Kafka ACL.

**Three possible solutions:**

Caveats: Can't update, so it will be a create and delete only action. 

1. Try the UUID on the k8s side.

Example:
Managed resources:
1. `name: marshmallory-abcd`
2. `name: marshmallory-defg`

Kafka side resources:
1. kafname: `marshmallory`
2. kafname: `marshmallory`

So kafname 1 and 2 appear the same, but have different underlying configurations such as a different user name, or different permissions.

In order to verify that we have the *right* ACL, it has to be described given ALL of the properties.

Pros: 
-Seems like it may work
Cons: 
- I have a feeling this will be a performance issue on large clusters.  This is mostly why I wanted to add Travis, and our Kafka folks who maybe can answer this.  
-We could potentially end up returning multiple ACLs if we don't search on EXACT terms.
-We will need to compare the k8s configuration constantly to make sure that if there was a change to the YAML that it then gets picked up in k8s/crossplane, and goes and does the search.  When it realizes there's a discrepancy, it has to be a delete and a recreate.
- This plan doesn't work at all like topics, because you can grab those by their unique name.

2. Figure out a way to map one k8s managed ACL to one Kafka ACL
Pros:
- This would ideally work the best, since we have two unique IDs to map
Cons:
- May have to actually manage a map of some kind
- Could be massive
- Only possible if there is some other way to get ACLs by ID or reference them directly.

3. Don't do this, at all.
I don't think there's pros to this other than writing something bad, and the cons are quite long.  This might be an option if there's a change in libraries or tech that would help this.  We can revisit.

TL;DR:
We are using the same name for each ACL, K8s overwrites whatever is there (except it can't because there's no update option), so it ends up creating a new one and leaving orphaned ACLs
We're definitely concerned over performance if k8s has to constantly hit the describe ACL, but I want to verify that with Travis because it is purely an assumption.
We want to make sure that these providers are written in a way that they should work, and maintain that relationship between k8s managed resources and the Kafka brokers.
We're afraid that by describing the ACLs constantly, that we may accidentally clobber other ACLs in the process, when the K8s manifests change.  I can elaborate if need be.

Happy for any and all thoughts, but we're a little stuck.

Reply:
@marshmallory I feel like 2nd solution could be possible.

AFAIU, there is no "id" or "name" of an ACL on the Kafka side.  So, to uniquely identify an ACL, we need to use all the fields in spec that build the ACL rule. What I am proposing is, let's use a custom external name (instead of using the name of the k8s resource) which is calculated using all fields like `User:Ed|ALLOW|ALTER_CONFIGS|TOPIC|marshmallory2|LITERAL|*` or even better as a `json` string like `{"PRINCIPAL":"User:Ed", "PERMISSION", "ALLOW", ...}`.

And use a flow like: 

![image](https://user-images.githubusercontent.com/9900707/153077048-75fa91b2-9ed6-4b61-9d17-f2449e4702b3.png)

Reply:
Ah I forgot to add checking that acl really exist on Kafka side at the end of Observe flow, and go to create if not. This would re-create rule if deleted from Kafka.

Reply:
Got it.  I understand this now.  Thanks for the diagram.

I think I have some questions, but I'll revisit with Ed and Jon and ask them as needed. Thanks :)

Reply:
Digging into the code now.  

I now fully understand the flow of how the name is set, but I am unsure of how to solve the problem of the name being set by default.

I see that it hits the reconciler, and that's where it creates the struct for the managed object (or something along those lines):

```go
	r := managed.NewReconciler(mgr,
		resource.ManagedKind(v1alpha1.AccessControlListGroupVersionKind),
		managed.WithExternalConnecter(&connector{
			kube:         mgr.GetClient(),
			usage:        resource.NewProviderConfigUsageTracker(mgr.GetClient(), &apisv1alpha1.ProviderConfigUsage{}),
			newServiceFn: kafka.NewAdminClient}),
		managed.WithLogger(l.WithValues("controller", name)),
		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))))
```
Where NewReconciler is a struct with managed in it:
```go
	r := &Reconciler{
		client:       m.GetClient(),
		newManaged:   nm,
		pollInterval: defaultpollInterval,
		timeout:      reconcileTimeout,
		managed:      defaultMRManaged(m),
		external:     defaultMRExternal(),
		log:          logging.NewNopLogger(),
		record:       event.NewNopRecorder(),
	}
```
Where defaultMRManaged has the NewNameAsExternalName within:
```go
func defaultMRManaged(m manager.Manager) mrManaged {
	return mrManaged{
		ConnectionPublisher: NewAPISecretPublisher(m.GetClient(), m.GetScheme()),
		Finalizer:           resource.NewAPIFinalizer(m.GetClient(), managedFinalizerName),
		Initializer:         NewNameAsExternalName(m.GetClient()),
		ReferenceResolver:   NewAPISimpleReferenceResolver(m.GetClient()),
	}
}
```
And finally it resets the name:
```go
// NewNameAsExternalName returns a new NameAsExternalName.
func NewNameAsExternalName(c client.Client) *NameAsExternalName {
	return &NameAsExternalName{client: c}
}
```

Do I override the name after the fact, and consider the `Name=ExternalName` to be "unset" if it doesn't fit our new generator?

I was expecting it to be unset totally (or to be able to do that), but then I realized that crossplane programmatically sets the external-name via the code above.

I could override it, I assume, with the Get/Set methods available in the meta.go, but that means the original name or an "unset" name would still always be Name=ExternalName

Alternatively, if there's another way to override it, that would be great.

Overriding it in the annotation doesn't do much, because it is just pulling from the metadata name.

I have a few ideas, including adding an extra field, but I'd love to know your thoughts.

Reply:
I should add, I did attempt to override it by directly setting a new external name via `SetExternalName()`, but it immediately gets overwritten and makes the name "marshmallory" again.

Reply:
I think you should disable initializer as here:

Reply:
Afk now, can check further if it doesn't help.

Btw, It is great that you've used this as an opportunity to dive into the crossplane runtime 🙌

Reply:
Please continue your AFK!

Just typing this out for tomorrow :)

https://github.com/crossplane/provider-aws/blob/b3f7c2b86595f053bf9295cd05dfac5a71d1535a/pkg/controller/servicediscovery/httpnamespace/setup.go#L60

Are you referring to the `managed.WithInitializers()` function in the above link?

Our setup function:
```go
func Setup(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter) error {
	name := managed.ControllerName(v1alpha1.AccessControlListGroupKind)

	o := controller.Options{
		RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
	}

	r := managed.NewReconciler(mgr,
		resource.ManagedKind(v1alpha1.AccessControlListGroupVersionKind),
		managed.WithExternalConnecter(&connector{
			kube:         mgr.GetClient(),
			usage:        resource.NewProviderConfigUsageTracker(mgr.GetClient(), &apisv1alpha1.ProviderConfigUsage{}),
			newServiceFn: kafka.NewAdminClient}),
		managed.WithLogger(l.WithValues("controller", name)),
		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))),
		managed.WithInitializers())

	return ctrl.NewControllerManagedBy(mgr).
		Named(name).
		WithOptions(o).
		For(&v1alpha1.AccessControlList{}).
		Complete(r)
}
```

I definitely don't think I have the usage or customization correct, but I couldn't find any docs.

Reply:
Yes @marshmallory looking correct 👍 

[It](https://github.com/crossplane/crossplane-runtime/blob/428b7c3903756bb0dcf5330f40298e1fa0c34301/pkg/reconciler/managed/reconciler.go#L556) will override the [default `Inializer` which contains NameAsExternalName](https://github.com/crossplane/crossplane-runtime/blob/428b7c3903756bb0dcf5330f40298e1fa0c34301/pkg/reconciler/managed/reconciler.go#L468) which would disable setting name as external name during [resource initialization](https://github.com/crossplane/crossplane-runtime/blob/428b7c3903756bb0dcf5330f40298e1fa0c34301/pkg/reconciler/managed/reconciler.go#L695).